code/modules/mob/dead/new_player/latejoin_menu.dm

```diff
@@ -0,0 +1,192 @@
+#define JOB_CHOICE_REROLL "Reroll"
+#define JOB_CHOICE_CANCEL "Cancel"
+
+GLOBAL_DATUM_INIT(latejoin_menu, /datum/latejoin_menu, new)
+
+/// Makes a list of jobs and pushes them to a DM list selector. Just in case someone did a special kind of fucky-wucky with TGUI.
+/datum/latejoin_menu/proc/fallback_ui(mob/dead/new_player/user)
+	var/list/jobs = list()
+	for(var/datum/job/job in SSjob.joinable_occupations)
+		jobs += job.title
+
+	var/input_contents = input(user, "Pick a job to join as:", "Latejoin Job Selection") as null|anything in jobs
+
+	if(!input_contents)
+		return
+
+	user.AttemptLateSpawn(input_contents)
+
+/datum/latejoin_menu/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "JobSelection", "Latejoin Menu")
+		ui.open()
+
+/datum/latejoin_menu/ui_data(mob/user)
+	var/mob/dead/new_player/owner = user
+	var/list/departments = list()
+	var/list/data = list(
+		"disable_jobs_for_non_observers" = SSlag_switch.measures[DISABLE_NON_OBSJOBS],
+		"round_duration" = DisplayTimeText(world.time - SSticker.round_start_time),
+		"departments" = departments,
+	)
+	if(SSshuttle.emergency)
+		switch(SSshuttle.emergency.mode)
+			if(SHUTTLE_ESCAPE)
+				data["shuttle_status"] = "The station has been evacuated."
+			if(SHUTTLE_CALL)
+				if(!SSshuttle.canRecall())
+					data["shuttle_status"] = "The station is currently undergoing evacuation procedures."
+
+	for(var/datum/job/prioritized_job in SSjob.prioritized_jobs)
+		if(prioritized_job.current_positions >= prioritized_job.total_positions)
+			SSjob.prioritized_jobs -= prioritized_job
+
+	for(var/datum/job_department/department as anything in SSjob.joinable_departments)
+		var/list/department_jobs = list()
+		var/list/department_data = list(
+			"jobs" = department_jobs,
+			"open_slots" = 0,
+		)
+		departments[department.department_name] = department_data
+
+		for(var/datum/job/job_datum as anything in department.department_jobs)
+			var/job_availability = owner.IsJobUnavailable(job_datum.title, latejoin = TRUE)
+
+			var/list/job_data = list(
+				"prioritized" = (job_datum in SSjob.prioritized_jobs),
+				"used_slots" = job_datum.current_positions,
+				"open_slots" = job_datum.total_positions < 0 ? "∞" : job_datum.total_positions,
+			)
+
+			if(job_availability != JOB_AVAILABLE)
+				job_data["unavailable_reason"] = get_job_unavailable_error_message(job_availability, job_datum.title)
+
+			if(job_datum.total_positions < 0)
+				department_data["open_slots"] = "∞"
+
+			if(department_data["open_slots"] != "∞")
+				department_data["open_slots"] += job_datum.total_positions - job_datum.current_positions
+
+			department_jobs[job_datum.title] = job_data
+
+	return data
+
+/datum/latejoin_menu/ui_static_data(mob/user)
+	var/list/departments = list()
+
+	for(var/datum/job_department/department as anything in SSjob.joinable_departments)
+		var/list/department_jobs = list()
+		var/list/department_data = list(
```

Comment:
You do not have to do it but this really should be constant data instead of static data, where it sends as a single json. It's trickier to setup the tgui though.

Reply:
Yea, if said constant data is JS/CSS-sided, then I'd really, really rather not, cause it just separates a bunch of related code for no obvious gain. Also makes it an absolute pain in the ass for extensibility.

Reply:
The obvious gain is that this is an enormous amount of data to send and slow to compute and send

Reply:
And it wouldn't be separate code--constant data means putting it into an asset file and then doing an async fetch in the menu